validator/src/main.rs

```diff
@@ -0,0 +1,216 @@
+        .iter()
+        .filter_map(|log| {
+            if log.starts_with("Program data: ") {
+                Some(log.strip_prefix("Program data: ").unwrap())
+            } else {
+                None
+            }
+        })
+        .collect();
+    let events: Vec<solana_ibc::events::NewBlock> = serialized_events
+        .iter()
+        .filter_map(|event| {
+            let decoded_event =
+                base64::prelude::BASE64_STANDARD.decode(event).unwrap();
+            let decoded_event: solana_ibc::events::Event =
+                borsh::BorshDeserialize::try_from_slice(&decoded_event)
+                    .unwrap();
+            match decoded_event {
+                solana_ibc::events::Event::NewBlock(e) => Some(e),
+                _ => {
+                    // println!("This is other event");
+                    None
+                }
+            }
+        })
+        .collect();
+    events
+}
+
+/// Solana sdk only accepts a keypair to form ed25519 instruction.
+/// Until they implement a method which accepts a pubkey and signature instead of keypair
+/// we have to use the below method instead.
+///
+/// Reference: https://github.com/solana-labs/solana/pull/32806
+pub fn new_ed25519_instruction_with_signature(
+    pubkey: &[u8],
+    signature: &[u8],
+    message: &[u8],
+) -> Instruction {
+    assert_eq!(pubkey.len(), PUBKEY_SERIALIZED_SIZE);
+    assert_eq!(signature.len(), SIGNATURE_SERIALIZED_SIZE);
+
+    let mut instruction_data = Vec::with_capacity(
+        DATA_START
+            .saturating_add(SIGNATURE_SERIALIZED_SIZE)
+            .saturating_add(PUBKEY_SERIALIZED_SIZE)
+            .saturating_add(message.len()),
+    );
+
+    let num_signatures: u8 = 1;
+    let public_key_offset = DATA_START;
+    let signature_offset =
+        public_key_offset.saturating_add(PUBKEY_SERIALIZED_SIZE);
+    let message_data_offset =
+        signature_offset.saturating_add(SIGNATURE_SERIALIZED_SIZE);
+
+    // add padding byte so that offset structure is aligned
+    instruction_data.extend_from_slice(bytes_of(&[num_signatures, 0]));
+
+    let offsets = solana_ibc::ed25519::SignatureOffsets {
+        signature_offset: signature_offset as u16,
+        signature_instruction_index: u16::MAX,
+        public_key_offset: public_key_offset as u16,
+        public_key_instruction_index: u16::MAX,
+        message_data_offset: message_data_offset as u16,
+        message_data_size: message.len() as u16,
+        message_instruction_index: u16::MAX,
+    };
+
+    instruction_data.extend_from_slice(bytes_of(&offsets));
+
+    debug_assert_eq!(instruction_data.len(), public_key_offset);
+
+    instruction_data.extend_from_slice(pubkey);
+
+    debug_assert_eq!(instruction_data.len(), signature_offset);
+
+    instruction_data.extend_from_slice(signature);
+
+    debug_assert_eq!(instruction_data.len(), message_data_offset);
+
+    instruction_data.extend_from_slice(message);
```

Comment:
I think this can be done simpler with concat:

```suggestion
    let instruction = [
        &[num_signatures, 0],
        bytes_of(&offsets),
        pubkey,
        signature,
        message
    ].concat()
```